### PR TITLE
Parse default unsafe fn

### PR DESCRIPTION
**Pull request by Avi-D-coder**

### CODE REVIEW

File: crates/ra_parser/src/grammar/items.rs [hunk 1]
```diff
@@ -121,7 +121,13 @@ pub(super) fn maybe_item(p: &mut Parser, m: Marker, flavor: ItemFlavor) -> Resul
...
                 // test default_unsafe_impl
                 // default unsafe impl Foo {}
-                if p.nth(2) == T![impl] {
+
+                // test default_unsafe_fn
+                // impl T for Foo {
+                //     default unsafe fn foo() {}
+                // }
+                let sk = p.nth(2);
+                if sk == T![impl] || sk == T![fn] {
```
Comment (matklad) [inline crates/ra_parser/src/grammar/items.rs:130]:
```suggestion
                if p.nth(2) == T![impl] || p.nth(2) == T![fn] {
```

We generally don't extract calls to nth.

Comment (matklad) [reply crates/ra_parser/src/grammar/items.rs:130]: Let's hope I edited the code correctly...

Comment (lnicola) [reply crates/ra_parser/src/grammar/items.rs]: > We generally don't extract calls to nth.

Do we have true lookahead here? I remember some quadratic behaviour around MBE expansion and parsing.

Comment (matklad) [reply crates/ra_parser/src/grammar/items.rs]: I think this should be constant-time yes. At some point, we should actually profile the stuff and make sure that the parser's code is tight, but, even than, I hope we'll convince the optimizer to CSE lokaheads...

Comment (lnicola) [reply crates/ra_parser/src/grammar/items.rs]: I was thinking of #1648 and #1643.

**Pull request merged**